pkg/limits/limits.go

```diff
@@ -17,4 +22,68 @@ const (
 	TTLMax                = 30 * time.Minute
 	ClusterAffiliatesMax  = 1024
 	AffiliateEndpointsMax = 64
+	RequestsPerSecondMax  = 5
 )
+
+// IPRateLimiter type.
+type IPRateLimiter struct {
+	IPs        map[string]*rate.Limiter
+	Mu         *sync.RWMutex
+	RateLimit  rate.Limit
+	BucketSize int
+}
+
+// NewIPRateLimiter returns newly initialized IPRateLimiter.
+func NewIPRateLimiter(rateLimit rate.Limit, bucketSize int) *IPRateLimiter {
+	i := &IPRateLimiter{
+		IPs:        make(map[string]*rate.Limiter),
+		Mu:         &sync.RWMutex{},
+		RateLimit:  rateLimit,
+		BucketSize: bucketSize,
+	}
+
+	return i
+}
+
+// AddIP creates a new rate limiter and adds it to the ips map,
+// using the IP address as the key.
+func (iPRL *IPRateLimiter) AddIP(ip string) *rate.Limiter {
+	iPRL.Mu.Lock()
+	defer iPRL.Mu.Unlock()
+
+	limiter := rate.NewLimiter(iPRL.RateLimit, iPRL.BucketSize)
+
+	iPRL.IPs[ip] = limiter
+
+	return limiter
+}
+
+// GetLimiter returns the rate limiter for the provided IP address if it exists.
+// Otherwise calls AddIP to add IP address to the map.
+func (iPRL *IPRateLimiter) GetLimiter(ip string) *rate.Limiter {
+	iPRL.Mu.Lock()
+	limiter, exists := iPRL.IPs[ip]
+
+	if !exists {
+		iPRL.Mu.Unlock()
+
+		return iPRL.AddIP(ip)
+	}
+
+	iPRL.Mu.Unlock()
+
+	return limiter
+}
+
+// IPAddrGarbageCollector periodically empties requests limit.
+func (iPRL *IPRateLimiter) IPAddrGarbageCollector() {
+	ticker := time.NewTicker(500 * time.Millisecond)
+
+	for range ticker.C {
+		for key, val := range iPRL.IPs {
+			if !val.Allow() {
```

Comment:
I feel this is the reverse of what we need, if `Allow()` is false, it means that the limiter is out of tokens, so the limit has been reached? so we shouldn't delete anything.

I believe proper condition should be something like `if AllowN(time.Now(), ipRL.BurstSize) {`